scripts/test/create-cluster.sh

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env bash
+
+set -eo pipefail
+
+TEST_DIR=$(cd "$(dirname "$0")" || exit 1; pwd)
+TEMPLATE_FILE="$TEST_DIR/template/eksctl/eks-cluster.yaml"
+
+source "$PWD/scripts/lib/common.sh"
+
+USAGE=$(cat << 'EOM'
+Usage: create-cluster.sh -n [cluster-name]
+Creates an EKS Cluster with Nodegroups. The Cluster is
+created using eksctl with a pre-defined eksctl template.
+ Required:
+ -n   Name of the EKS cluster
+ Optional:
+ -v   K8s Version. Defaults to 1.20
+ -r   Region of the EKS Cluster. Defaults to us-west-2
+ -c   eksctl Cluster config path. Defaults to build/eks-cluster.yaml
+ -t   Type of cluster. Defaults to eksctl cluster. Can be set to "kops" or "eksctl"
+EOM
+)
+
+
+
+while getopts "n:v:r:c:t:" o; do
+  case "${o}" in
+    n) # Name of the EKS Cluster
+      CLUSTER_NAME=${OPTARG}
+      ;;
+    v) # K8s Version of the EKS Cluster
+      K8S_VERSION=${OPTARG}
+      ;;
+    r) # Region where EKS Cluster will be created
+      AWS_REGION=${OPTARG}
+      ;;
+    c) # eksctl cluster config file path
+      CLUSTER_CONFIG_PATH=${OPTARG}
+      ;;
+    t) # Type of cluster to be created
+      CLUSTER_TYPE=${OPTARG}
+      ;;
+    *)
+      echoerr "${USAGE}"
+      exit 1
+      ;;
+  esac
+done
+
+
+
+shift $((OPTIND-1))
+
+if [[ -z "$CLUSTER_NAME" ]]; then
+ echoerr "${USAGE}\n\nmissing: -n is a required flag\n"
+ exit 1
+fi
+
+if [[ -z "$K8S_VERSION" ]]; then
+  K8S_VERSION="1.20"
+  echo "no k8s version defined, will fallback to default version $K8S_VERSION"
+fi
+
+if [[ -z "$AWS_REGION" ]]; then
+  AWS_REGION="us-west-2"
+  echo "no regions defined, will fallback to default region $AWS_REGION"
+fi
+
+if [[ -z "$CLUSTER_CONFIG_PATH" ]]; then
+  CLUSTER_CONFIG_DIR="$TEST_DIR/build"
+  CLUSTER_CONFIG_PATH="$CLUSTER_CONFIG_DIR/eks-cluster.yaml"
+
+  echo "cluster config path not defined, will generate it at default path $CLUSTER_CONFIG_PATH"
+
+  mkdir -p "$CLUSTER_CONFIG_DIR"
+fi
+
+
+function create_eksctl_cluster() {
+  check_is_installed eksctl
+  echo "creating a $K8S_VERSION cluster named $CLUSTER_NAME using eksctl"
+
+  # Using a static Instance Role Name so IAM Policies
+  # can be attached to it later without having to retrieve
+  # the auto generated eksctl name.
+  sed "s/CLUSTER_NAME/$CLUSTER_NAME/g;
+  s/CLUSTER_REGION/$AWS_REGION/g;
+  s/K8S_VERSION/$K8S_VERSION/g;
+  s/INSTANCE_ROLE_NAME/$INSTANCE_ROLE_NAME/g" \
+  "$TEMPLATE_FILE" > "$CLUSTER_CONFIG_PATH"
+
+  eksctl create cluster -f "$CLUSTER_CONFIG_PATH"
+}
+
+
+
+function create_kops_cluster {
```

Comment:
Would be ideal if we could have a shared library which can be reused by VPC CNI, VPC RC, ALB and other networking components for common functions like create cluster (EKS/KOPS) and delete cluster (EKS/KOPS)

Reply:
I definitely believe that it will ensure consistency across repos since each repo has it’s own version and it will be good from point of future extensibility. But I just concerned how much time it will take considering that cluster creation/deletion may be used in number of scripts across repos and will require considerable amount of  testing to make sure nothing is broken. Also, some of the create delete scripts are tightly coupled with other components or for instance some of the ENV variables for ALB cluster creation/deletion script maybe different from ones used in vpc-resource controller cluster creation/deletion scripts. Besides, I noticed that other repos don’t use kops in addition for cluster/creation deletion and is only used by cni.